### ISSUE TITLE
Support 802.1Q tags

### ISSUE THREAD

**Issue by phoracek** [libpnet/libpnet#172]
VLAN tags handling is usually done on switches level and not needed on end-points. This is not true for virtual networking hypervisors which could be connected via trunk in order to divide multiple VM networks.

`EthernetPacket` should have one more method `fn get_tag(&self) -> Option(VlanTag)`

What has to be done to implement this? I guess it is possible even now, match 0x8100 EtherType and then handle tag and real EtherType in payload. But native support would be great.

**Comment by mrmonday** [comment]
Hi there!

Sorry I missed you on IRC earlier. There are quite a few ways this could be implemented - I don't know which is the nicest. Perhaps something like:

```rust
#[packet]
pub struct Vlan {
    priority_code_point: <either u3 or an enum>,
    drop_eligible_indicator: u1,
    vlan_identifier: u12be,
    #[construct_with(u16)]
    ethertype: EtherType,
    #[payload]
    payload: Vec<u8>,
}
```

Would work?

Then you can match on the ethertype of the original Ethernet packet, use `VlanPacket` as the payload, then match on the ethertype of the Vlan packet for the rest of the packet.

I'd be happy to mentor you if you're interested in implementing this.

**Comment by phoracek** [comment]
Yea, that looks good, thanks! However, it adds quite a bit overhead (or at least I'm not able to think about an elegant handling).

I'd like to have something like this:
```rust
match {
    EthernetPacket(destination, source, ethertype, payload) => ...,
    VlanPacket(destination, source, vlanstuff, ..., ethertype, payload) => ...
}
```
I don't know if it is possible to implement it. Can we somehow decide whether to use one or another structure based on some further handling (check EtherType and decide).

I'd like to implement this functionality (802.1Q Packet and matching) into `libpnet`, but I definitely need your help.

**Comment by mrmonday** [comment]
Unfortunately there's no clean/easy way to match on packets with libpnet yet - it's something I'm hoping to have support for by version 1.0 (this is part of issue #139). The packet dump example in the examples directory shows how you have to write this currently.

In terms of a more terse way of writing it, I doubt it will be possible to use `match`, but it may be possible to write a macro which provides a similar syntax. I haven't given much thought about how best to implement it - ideas are welcome if it is something that interests you!

In terms of adding support for the packet to libpnet - take a look in `src/packet/` at how the other packets are implemented. There's a generic wrapper file for each packet (there's no need to understand what it's doing, unless you're particularly interested), and an actual implementation file (the `.in.rs`). You'll also need to [update the `FILES` array in build.rs](https://github.com/libpnet/libpnet/blob/f08e08e76e6dc7498171982f58caa7b2c1270b89/build.rs#L28-L33) to make sure the new file gets built.

If you have any questions about how it all fits together (or anything at all really), just ask!

**Comment by moosingin3space** [comment]
Why is this issue still open?

**Comment by TheRealNOIG** [comment]
I would like to reopen this issue as I believe the current solution has some issues.

As mentioned in the IRC channel I think there are multiple points on why the current implementation may cause issues and/or be structured in an inefficient way.

Firstly, the current implementation segregates frames that have a . 1Q header's in their frame to a new Ethernet packet struct.
However this dosnt make much sense as it is not a protocol itself but an option that modifies the frame of all Ethernet frames.
Because of this the current implementation would require the dev to check not only the 'normal' packet type but also if there are any vlan packets and then perform the packet type check again.
This would require 3 checks just to make sure you always get the frame you want.

Soundly, this may also cause a 'bug' or unintended side effect to any program that was not written to be "VLAN awars"
This would be due to some networking equipment creating .1q header on all packets that lack them when they enter that devices VLAN aware network.
This would cause any program that only checks the protocol eth type instead of both the protocol and the VLAN version if it where placed on a vlan aware network with a vlan aware interface as now all packets coming into that interface would be made into VLANPackets.

For example the examples/packetdump.rs example probably wont work on vlan aware interface (Cant test this till I get home tonight)
https://github.com/libpnet/libpnet/blob/main/examples/packetdump.rs


To recap, the current structure dosnt make much sense as all Ethernet frames can have a .1q header no mater what the protocol is.
Because of this it dosnt make much sense to have a separate Ethernet type that then holds the real ethertype inside itself.
I think using the original suggestion of adding a vlan option to all ethenet packets would make more sense.
Now it should probably be the full .1q header instead of just the vlan tag as originally suggested.
